pyvista/core/dataset.py

```diff
@@ -2166,6 +2166,69 @@ def find_closest_cell(self, point: Union[int, np.ndarray]) -> Union[int, np.ndar
             closest_cells.append(int(cellId))
         return closest_cells[0] if len(closest_cells) == 1 else np.array(closest_cells)
 
+    def find_containing_cell(self, point: Union[int, np.ndarray]) -> Union[int, np.ndarray]:
+        """Find index of a cell that contains the given point.
+
+        Parameters
+        ----------
+        point : iterable(float) or np.ndarray
+            Coordinates of point to query (length 3) or a ``numpy`` array of ``n``
+            points with shape ``(n, 3)``.
+
+        Returns
+        -------
+        int or numpy.ndarray
+            Index or indices of the cell in this mesh that contains
+            the given point.
+
+        Examples
+        --------
+        A unit square with 16 equal sized cells is created and a cell
+        containing the point ``[0.3, 0.3, 0.0]`` is found.
+
+        >>> import pyvista
+        >>> mesh = pyvista.UniformGrid(dims=[5, 5, 1], spacing=[1/4, 1/4, 0])
+        >>> mesh # doctest:+ELLIPSIS
+        UniformGrid...
+        >>> mesh.find_containing_cell([0.3, 0.3, 0.0])
+        5
+
+        A point outside the mesh domain will return ``-1``.
+
+        >>> mesh.find_containing_cell([0.3, 0.3, 1.0])
+        -1
+
+        Find the cells that contain 1000 random points inside the mesh.
+        >>> import numpy as np
+        >>> points = np.random.random((1000, 3))
+        >>> indices = mesh.find_containing_cell(points)
+        >>> indices.shape
+        (1000,)
+
+        """
+        if isinstance(point, collections.abc.Sequence):
+            point = np.array(point)
+        # check if this is an array of points
+        if isinstance(point, np.ndarray):
+            if point.ndim > 2:
+                raise ValueError("Array of points must be 2D")
+            if point.ndim == 2:
+                if point.shape[1] != 3:
+                    raise ValueError("Array of points must have three values per point")
+            else:
+                if point.size != 3:
+                    raise ValueError("Given point must have three values")
+                point = np.array([point])
```

Comment:
Just nitpick (this will not copy):
```suggestion
                point = point.reshape(1, 3)
```

Reply:
Since point may be mutable, don't we want to copy to avoid mutating the input arg?

Reply:
If you just mean the reshape here: it's not in-place. This will rebind a view of the input array to the name `point`.

If you mean potentially mutating this array later in the method: in that case we also have to make a copy on the non-1d-input path.

Reply:
I see most of your first point now, although I'm not sure I've grasped all of the nuance in all cases.  Since we feed this vector into vtk, it is a bit of a black box.  This method in `vtkCellLocator` does modify some of the inputs, although I'm mostly certain `point` is not modified.  I'm not sure if we should trust vtk in general for all uses (see my next point).  There is a downside to copying too due to increase in memory requirement.

This has to be a very common use case throughout the codebase, where we want to allow either an arg to be a single sequence of length 3 or an ndarray of shape (n, 3).  Then the code checks this and standardizes to a (n, 3) ndarray.  This would make a good helper function.

It could be something like the below where it is optional to make a copy.  Then we could use `copy=False` when we know that it is not modified or `copy=True` when we know it is modified.  The default is True as it is the safest.

```py
def _coerce_arg_shape_n_3(array, copy=True):
    ...
```

I think this should be a separate PR as it could touch many parts of the code.

Reply:
To rephrase my point (no pun intended): `point = point.reshape(-1, 3)` won't change the shape of the input array but it will be a view into it, meaning that it will propagate (value) mutation to the input arrays. So I think you understood my point :)

And your point about the helper is excellent, I think what you laid out is a good plan, and I agree it should be a separate PR.

Reply:
Just so I don't forget later: perhaps we could try coming up with tests that try to verify that inputs are not mutated in these various filters. Assuming we think that mutation of certain inputs is not dependent on a rare constellation of parameters.

Reply:
I think this has cemented my understanding.  Thanks!

I'm going to make this an issue as a maintenance item so we don't lose track of it and resolve these conversations.